DBZ-5902: Add docs for PartitionRouting SMT that will replace ComputePartition (#4410)
debezium/debezium: documentation/modules/ROOT/pages/transformations/partition-routing.adoc

Based on the configuration, the SMT routes change events for the records that have the field name `hammer` to the same partition.

> That is, the items with `id` values `104`, `105`, and `106` are routed to the same partition.

Author: @roldanbob ping me once available for a call.

In case you have defined same number of partitions for al topics, this configuration

[source]
----
...
topic.creation.default.partitions=2
topic.creation.default.replication.factor=1
...
topic.prefix=fulfillment
transforms=ComputePartition
transforms.ComputePartition.type=io.debezium.transforms.partitions.ComputePartition
transforms.ComputePartition.partition.data-collections.field.mappings=inventory.products:name,inventory.orders:purchaser
transforms.ComputePartition.partition.data-collections.partition.num.mappings=inventory.products:2,inventory.orders:2
...
----

must be replaced with this

[source]
----
...
topic.creation.default.partitions=2
topic.creation.default.replication.factor=1
...

topic.prefix=fulfillment
transforms=PartitionRouting
transforms.PartitionRouting.type=io.debezium.transforms.partitions.PartitionRouting
transforms.PartitionRouting.partition.payload.field=change.name,change.purchaser
transforms.PartitionRouting.partition.topic.num=2
transforms.PartitionRouting.predicate=allTopic
predicates=allTopic
predicates.allTopic.type=org.apache.kafka.connect.transforms.predicates.TopicNameMatches
predicates.allTopic.pattern=fulfillment.*
...
----

> In case you have different partition number for topics, for example, the relative topic for data collection `products` has been configured with 3 partitions and
> for the data collection `orders` has been configured with 2 partitions.
> 
> The following old configuration

Reviewer: ```suggestion
If the SMT emits events to topics that do not share the same number of partitions, you must specify unique `partition.num.mappings` values for each topic.
For example, in the following example, the topic for the legacy `products` collection is configured with 3 partitions, and the topic for the `orders` data collection is configured with 2 partitions: 

.Example: Legacy `ComputePartition` configuration that sets unique partition values for different topics 
```

[source]
----
...
topic.prefix=fulfillment
transforms=ComputePartition
transforms.ComputePartition.type=io.debezium.transforms.partitions.ComputePartition
transforms.ComputePartition.partition.data-collections.field.mappings=inventory.products:name,inventory.orders:purchaser
transforms.ComputePartition.partition.data-collections.partition.num.mappings=inventory.products:3,inventory.orders:2
...
----

> must be replaced with this new configuration

Reviewer: ```suggestion
Replace the preceding `ComputePartition` configuration with the following `PartitionRouting` configuration:
```

> 

Reviewer: ```suggestion
.`PartitionRouting` configuration that sets unique `partition.topic.num` values for different topics 
```

Based on the configuration, the SMT routes change events for the records that have the field name `hammer` to the same partition.
That is, the items with `id` values `104`, `105`, and `106` are routed to the same partition.

// Type: concept
// Title: Example: Advanced configuration of the {prodname} partition routing SMT
// ModuleID: advanced-configuration-of-the-debezium-partition-routing-smt
[[example-advanced-partition-routing-configuration-example]]
== Example: Advanced configuration

Suppose you have different data collections (t1, t2) routed to same topic (my_topic), and you want to partition events from data collection t1 using field f1
and events from data collection t2 using field f2.

> This will be the configuration

Reviewer: ```suggestion
You could apply the following configuration:
```

[source]
----
transforms=PartitionRouting
transforms.PartitionRouting.type=io.debezium.transforms.partitions.PartitionRouting
transforms.PartitionRouting.partition.payload.fields=change.f1,change.f2
transforms.PartitionRouting.partition.topic.num=2
transforms.PartitionRouting.predicate=myTopic

predicates=myTopic
predicates.myTopic.type=org.apache.kafka.connect.transforms.predicates.TopicNameMatches
predicates.myTopic.pattern=my_topic
----

> Note that in this configuration we omitted the part to configure the re-routes of events from different topic to a specific one. See {link-prefix}:{link-topic-routing}[Topic Routing SMT].

Reviewer: As much as possible, it's best to avoid sentences that begin with the construction "Note that ..." or similar. Place content that should be in a note in an asciidoc admonition (Note, Important, Warning, and so forth). Otherwise, rework the content into the main flow.

> The following table lists the configuration options that you can use with the partition routing SMT.

Reviewer: ```suggestion
The following table lists the configuration options that you can set for the partition routing SMT.
```

.Partition routing SMT (`PartitionRouting`) configuration options
[cols="30%a,25%a,45%a"]
|===
|Property
|Default
|Description

> |[[partition-routing-payload-fields]]<<partition-routing-payload-fields, `partition.payload.field`>>

Reviewer: Not a big deal, but I'm inclined to have the field name value match the string in the anchor URL. WDYT?

> |[[partition-routing-payload-fields]]<<partition-routing-payload-fields, `partition.payload.field`>>

Author: Yes, I agree but the right one is fields. I'll change the property name also on code. Thanks for catching it.

|

> |Payload fields to use to calculate the partition. Supports Struct nesting using dot notation. To access fields related to data collections, you can use: after, before or change,

Reviewer: ```suggestion
|Specifies the fields in the event payload that the SMT uses to calculate the target partition.
Use dot notation if you want the SMT to add fields from the original payload to specific levels in the output data structure. 
To access fields related to data collections, you can use: `after`, `before`, or `change`.
```

> where 'change' is a special field that will automatically choose, based on operation, the 'after' or 'before'. If a field not exist for the current record it will simply not used.

Reviewer: ```suggestion
The 'change' field is a special field that results in the SMT automatically populating content in the 'after' or 'before' elements, depending on type of operation. 
If a specified field is not present in a record, the SMT skips it.
```
